File: Seori/433. Minimum Genetic Mutation.py
class Solution:
    def minMutation(self, startGene: str, endGene: str, bank: List[str]) -> int:
        dict_mutate = {'A' : ('C', 'G', 'T'),
                     'C' : ('A', 'G', 'T'),
                     'G' : ('A', 'C', 'T'),
                     'T' : ('A', 'C', 'G')}

        visited = [startGene]
        queue = [(startGene, 0)]
        while queue:
            nowGene, count = queue.pop(0)
            if nowGene == endGene:
                return count
            for i in range(len(nowGene)):
                for mutate in dict_mutate:
                    nextGene = nowGene[:i] + mutate + nowGene[i+1:]
                    if nextGene in bank and nextGene not in visited:
                        queue.append((nextGene, count + 1))
                        visited.append(nextGene)
        return -1

File: Lucy/src/2025/2025-8-W02/433_minimum_genetic_mutation.ts
function minMutation(startGene: string, endGene: string, bank: string[]): number {
  if (!bank.includes(endGene)) return -1;

  const queue: [string, number][] = [[startGene, 0]];

  const visited = new Set<string>();
  visited.add(startGene);

  while (queue.length > 0) {
    const [currentGene, count] = queue.shift()!;

    if (currentGene === endGene) return count;

    for (const validGene of bank) {
      const mutationCount = Array.from(validGene).filter((ch, i) => ch !== currentGene[i]).length;
      if (mutationCount === 1 && !visited.has(validGene)) {
        visited.add(currentGene);

        queue.push([validGene, count + 1]);
      }
    }
  }

  return -1;
}
